File: sdk/java/src/main/java/org/finos/openresourcebroker/sdk/process/SubprocessManager.java
// Layer 1: Subprocess Manager
package org.finos.openresourcebroker.sdk.process;

import java.io.*;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.UnixDomainSocketAddress;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.channels.SocketChannel;
import java.nio.channels.Channels;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Logger;

/**
 * Manages the lifecycle of a local ORB subprocess.
 *
 * <p>Responsibilities:
 * <ul>
 *   <li>Spawn {@code orb server start --foreground --api-only [--socket-path <path>]}
 *   <li>Poll {@code /health} until healthy or timeout
 *   <li>Background health-check loop; mark unhealthy after N consecutive failures
 *   <li>Graceful stop: SIGTERM → wait → SIGKILL fallback
 * </ul>
 *
 * <p>Thread-safe; all state transitions use atomics.
 */
public class SubprocessManager {

    private static final Logger LOG = Logger.getLogger(SubprocessManager.class.getName());

    private static final Duration STARTUP_POLL_INTERVAL = Duration.ofMillis(200);
    private static final long BG_POLL_INTERVAL_S = 5;
    private static final Duration BG_POLL_TIMEOUT = Duration.ofSeconds(2);
    private static final int UNHEALTHY_THRESHOLD = 3;

    private final ProcessConfig config;
    private Process process;
    private final AtomicBoolean healthy = new AtomicBoolean(false);
    private final AtomicInteger consecutiveFail = new AtomicInteger(0);
    private ScheduledExecutorService scheduler;
    private volatile boolean stopped = false;

    public SubprocessManager(ProcessConfig config) {
        this.config = config;
    }

    /**
     * Start the ORB subprocess and wait for it to become healthy.
     *
     * @throws IOException if the process cannot be started
     * @throws InterruptedException if the calling thread is interrupted
     * @throws IllegalStateException if ORB does not become healthy within the timeout
     */
    public synchronized void start() throws IOException, InterruptedException {
        String binary = resolveBinary(config.getBinary());
        List<String> cmd = buildCommand(binary);

        LOG.info("Starting ORB: " + String.join(" ", cmd));

        ProcessBuilder pb = new ProcessBuilder(cmd);
        if (config.getEnv() != null) {
            pb.environment().putAll(config.getEnv());
        }
        // Redirect stderr to stdout so we can see errors
        pb.redirectErrorStream(false);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        pb.redirectError(ProcessBuilder.Redirect.DISCARD);

        process = pb.start();
        LOG.info("ORB process started, PID=" + process.pid());

        // Wait for healthy
        long deadline = System.currentTimeMillis() + config.getStartTimeout().toMillis();
        while (System.currentTimeMillis() < deadline) {
            Thread.sleep(STARTUP_POLL_INTERVAL.toMillis());
            if (!process.isAlive()) {
                throw new IllegalStateException("ORB process exited unexpectedly (exit=" +
                        process.exitValue() + ")");
            }
            if (pollHealth()) {
                healthy.set(true);
                LOG.info("ORB process healthy, PID=" + process.pid());
                startMonitor();
                return;
            }
        }

        process.destroyForcibly();
        throw new IllegalStateException(
                "ORB did not become healthy within " + config.getStartTimeout());
    }

    /**
     * Stop the managed process gracefully (SIGTERM then SIGKILL).
     */
    public synchronized void stop() {
        if (stopped) return;
        stopped = true;
        healthy.set(false);

        if (scheduler != null) {
            scheduler.shutdownNow();
        }

        if (process == null) return;

        // Snapshot descendants BEFORE terminating the parent, otherwise the
        // process tree is no longer walkable once the parent dies.
        List<ProcessHandle> descendants = process.descendants().collect(Collectors.toList());

        // SIGTERM the parent first, then its descendants, for a graceful stop.
        process.destroy();
        descendants.forEach(ProcessHandle::destroy);

        try {
            boolean exited = process.waitFor(
                    config.getStopTimeout().toSeconds(), TimeUnit.SECONDS);
            if (!exited) {
                // Escalate to SIGKILL on the parent and any surviving children so
                // no orphaned child process is left behind.
                process.destroyForcibly();
                descendants.forEach(ProcessHandle::destroyForcibly);
            } else {
                // Parent gone; force-kill any child that ignored SIGTERM.
                descendants.stream()
                        .filter(ProcessHandle::isAlive)
                        .forEach(ProcessHandle::destroyForcibly);
            }
        } catch (InterruptedException e) {
            process.destroyForcibly();
            descendants.forEach(ProcessHandle::destroyForcibly);
            Thread.currentThread().interrupt();
        }
    }

    /** Returns true if the managed process is currently healthy. */
    public boolean isHealthy() {
        return healthy.get();
    }

    /** Returns the PID of the managed process, or -1 if not started. */
    public long getPid() {
        return process != null ? process.pid() : -1L;
    }

    // ------------------------------------------------------------------
    // Internal
    // ------------------------------------------------------------------

    private List<String> buildCommand(String binary) {
        List<String> cmd = new ArrayList<>();
        cmd.add(binary);

        // Support python -m orb invocation
        if (binary.endsWith("python") || binary.endsWith("python3") || binary.endsWith("python3.exe")) {
            cmd.add("-m");
            cmd.add("orb");
        }

        // Base args
        if (config.getConfigPath() != null) {
            cmd.add("--config");
            cmd.add(config.getConfigPath());
        }
        cmd.add("server");
        cmd.add("start");
        cmd.add("--foreground");
        cmd.add("--api-only");

        if (config.getSocketPath() != null && !config.getSocketPath().isEmpty()) {
            cmd.add("--socket-path");
            cmd.add(config.getSocketPath());
        } else {
            cmd.add("--port");
            cmd.add(String.valueOf(config.getPort() > 0 ? config.getPort() : 8000));
        }

        if (config.getExtraArgs() != null) {
            cmd.addAll(config.getExtraArgs());
        }
        return cmd;
    }

    private String resolveBinary(String binary) {
        if (binary == null || binary.isEmpty()) binary = "orb";
        if (isOnPath(binary)) return binary;

        // Try python -m orb
        for (String py : new String[]{"python3", "python"}) {
            if (isOnPath(py)) return py;
        }
        throw new IllegalStateException(
                "Cannot find ORB binary '" + binary + "' in PATH and python/python3 not available");
    }

    private static boolean isOnPath(String binary) {
        String path = System.getenv("PATH");
        if (path == null) return false;
        for (String dir : path.split(File.pathSeparator)) {
            File f = new File(dir, binary);
            if (f.canExecute()) return true;
        }
        return false;
    }

    private boolean pollHealth() {
        try {
            String healthUrl;
            if (config.getSocketPath() != null && !config.getSocketPath().isEmpty()) {
                healthUrl = "http://localhost/health";
                return pollHealthUds(config.getSocketPath(), healthUrl);
            } else {
                int port = config.getPort() > 0 ? config.getPort() : 8000;
                healthUrl = "http://localhost:" + port + "/health";
                return pollHealthTcp(healthUrl);
            }
        } catch (Exception e) {
            return false;
        }
    }

    private boolean pollHealthUds(String socketPath, String url) {
        try {
            UnixDomainSocketAddress addr = UnixDomainSocketAddress.of(socketPath);
            try (SocketChannel ch = SocketChannel.open()) {
                ch.configureBlocking(true);
                // Set connect timeout via socket options
                try {
                    ch.connect(addr);
                } catch (IOException e) {
                    return false;
                }

                OutputStream out = Channels.newOutputStream(ch);
                InputStream in = Channels.newInputStream(ch);

                String req = "GET /health HTTP/1.1\r\nHost: localhost\r\nAccept: application/json\r\nConnection: close\r\n\r\n";
                out.write(req.getBytes(StandardCharsets.US_ASCII));
                out.flush();

                return parseHealthResponse(in);
            }
        } catch (Exception e) {
            return false;
        }
    }

    private boolean pollHealthTcp(String url) {
        try {
            URI uri = URI.create(url);
            java.net.Socket sock = new java.net.Socket();
            sock.connect(new InetSocketAddress(uri.getHost(),
                    uri.getPort() == -1 ? 80 : uri.getPort()), 2000);
            sock.setSoTimeout(2000);
            try (sock) {
                String req = "GET /health HTTP/1.1\r\nHost: " + uri.getHost() + "\r\nAccept: application/json\r\nConnection: close\r\n\r\n";
                sock.getOutputStream().write(req.getBytes(StandardCharsets.US_ASCII));
                sock.getOutputStream().flush();
                return parseHealthResponse(sock.getInputStream());
            }
        } catch (Exception e) {
            return false;
        }
    }

    private boolean parseHealthResponse(InputStream in) throws IOException {
        BufferedReader reader = new BufferedReader(
                new InputStreamReader(in, StandardCharsets.UTF_8));

        // Status line
        String statusLine = reader.readLine();
        if (statusLine == null) return false;
        String[] parts = statusLine.split(" ", 3);
        if (parts.length < 2) return false;
        int code;
        try { code = Integer.parseInt(parts[1].trim()); }
        catch (NumberFormatException e) { return false; }

        if (code == 401) return false; // auth configured on /health
        if (code != 200) return false;

        // Skip headers
        String line;
        while ((line = reader.readLine()) != null && !line.isEmpty()) { }

        // Body
        StringBuilder body = new StringBuilder();
        char[] buf = new char[1024];
        int n;
        while ((n = reader.read(buf)) != -1) body.append(buf, 0, n);

        String b = body.toString().trim();
        return b.contains("\"healthy\"") || b.contains("\"degraded\"");
    }

    private void startMonitor() {
        scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
            Thread t = new Thread(r, "orb-health-monitor");
            t.setDaemon(true);
            return t;
        });
        scheduler.scheduleAtFixedRate(() -> {
            if (stopped) return;
            if (pollHealth()) {
                consecutiveFail.set(0);
                if (!healthy.get()) {
                    healthy.set(true);
                    LOG.info("ORB process recovered");
                }
            } else {
                int n = consecutiveFail.incrementAndGet();
                if (n >= UNHEALTHY_THRESHOLD) {
                    healthy.set(false);
                    LOG.warning("ORB process marked unhealthy after " + n + " failures");
                }
            }
        }, BG_POLL_INTERVAL_S, BG_POLL_INTERVAL_S, TimeUnit.SECONDS);
    }
}

File: sdk/kotlin/src/main/kotlin/org/finos/openresourcebroker/sdk/process/SubprocessManager.kt
/**
 * Layer 1: Subprocess Manager
 *
 * Spawns and supervises a local ORB process, polls /health until healthy,
 * and stops it cleanly.
 *
 * Mirrors sdk/go/internal/process/manager.go behaviour:
 *   - Binary path + args + env
 *   - Poll /health via UDS or TCP until healthy or timeout
 *   - Background health-check loop: unhealthy after N consecutive failures
 *   - Graceful stop: SIGTERM -> wait -> SIGKILL fallback
 *   - Thread-safe via ReentrantLock + @Volatile
 */

package org.finos.openresourcebroker.sdk.process

import kotlinx.coroutines.*
import okhttp3.OkHttpClient
import okhttp3.Request
import org.finos.openresourcebroker.sdk.transport.UdsSocketFactory
import java.io.File
import java.time.Duration
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private const val STARTUP_POLL_INTERVAL_MS = 200L
private const val BG_POLL_INTERVAL_MS = 5_000L
private const val BG_POLL_TIMEOUT_MS = 2_000L
private const val UNHEALTHY_THRESHOLD = 3

/**
 * Configuration for spawning a managed ORB subprocess.
 */
data class ProcessConfig(
    /** Path to the orb binary (default: "orb"). Accepts "python3 -m orb" style too. */
    val binary: String = "orb",
    /** Extra arguments appended after "server start --foreground --api-only" */
    val args: List<String> = emptyList(),
    /** Additional environment variables merged with system env */
    val env: Map<String, String> = emptyMap(),
    /** UNIX socket path; if set, UDS transport is used for health checks */
    val socketPath: String = "",
    /** TCP port (ignored when socketPath is set) */
    val port: Int = 8000,
    /** Max ms to wait for /health to return healthy (default 45s) */
    val startTimeoutMs: Long = 45_000L,
    /** SIGTERM grace period before SIGKILL (default 10s) */
    val stopTimeoutMs: Long = 10_000L,
    /** PYTHONPATH injection (useful when running from source) */
    val pythonPath: String = "",
)

/**
 * Manages an ORB subprocess lifecycle.
 */
class SubprocessManager(private val cfg: ProcessConfig) {

    private val lock = ReentrantLock()
    private var proc: Process? = null
    private val _healthy = AtomicBoolean(false)
    private val consecutiveFail = AtomicInteger(0)
    private val stopped = AtomicBoolean(false)
    private var bgJob: Job? = null

    /** True if the managed process is currently healthy. */
    val healthy: Boolean get() = _healthy.get()

    /**
     * Start the ORB subprocess and wait for it to become healthy.
     * Throws [IllegalStateException] if already started.
     * Throws [RuntimeException] if the process does not become healthy within the timeout.
     */
    suspend fun start() {
        lock.withLock {
            check(proc == null) { "SubprocessManager: already started" }

            val command = buildCommand()
            val fullEnv = buildEnvironment()

            val pb = ProcessBuilder(command)
                .redirectErrorStream(false)
                .also { builder ->
                    builder.environment().putAll(fullEnv)
                }

            proc = pb.start()

            // Drain stdout/stderr to prevent blocking
            val process = proc!!
            Thread {
                try { process.inputStream.copyTo(java.io.OutputStream.nullOutputStream()) } catch (_: Exception) {}
            }.also { it.isDaemon = true; it.start() }
            Thread {
                try { process.errorStream.copyTo(java.io.OutputStream.nullOutputStream()) } catch (_: Exception) {}
            }.also { it.isDaemon = true; it.start() }
        }

        // Poll until healthy (outside lock to allow cancellation)
        val deadline = System.currentTimeMillis() + cfg.startTimeoutMs
        while (System.currentTimeMillis() < deadline) {
            delay(STARTUP_POLL_INTERVAL_MS)
            if (!isRunning()) {
                throw RuntimeException("SubprocessManager: ORB process exited prematurely")
            }
            if (pollHealth()) {
                _healthy.set(true)
                startBgMonitor()
                return
            }
        }

        kill()
        throw RuntimeException(
            "SubprocessManager: orb did not become healthy within ${cfg.startTimeoutMs}ms"
        )
    }

    /**
     * Stop the managed subprocess gracefully.
     * Sends SIGTERM, waits for stopTimeoutMs, then sends SIGKILL.
     */
    fun stop() {
        stopped.set(true)
        _healthy.set(false)
        bgJob?.cancel()
        bgJob = null

        val process = lock.withLock { proc } ?: return

        // Snapshot descendants BEFORE terminating the parent, otherwise the
        // process tree is no longer walkable once the parent dies.
        val descendants = process.descendants().toList()

        // SIGTERM the parent first, then its descendants, for a graceful stop.
        process.destroy()
        descendants.forEach { it.destroy() }

        val exited = process.waitFor(cfg.stopTimeoutMs, TimeUnit.MILLISECONDS)
        if (!exited) {
            // Escalate to SIGKILL on the parent and any surviving children so no
            // orphaned child process is left behind.
            process.destroyForcibly()
            descendants.forEach { it.destroyForcibly() }
            process.waitFor(5, TimeUnit.SECONDS)
        } else {
            // Parent gone; force-kill any child that ignored SIGTERM.
            descendants.filter { it.isAlive }.forEach { it.destroyForcibly() }
        }
    }

    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------

    private fun isRunning(): Boolean {
        val p = lock.withLock { proc } ?: return false
        return p.isAlive
    }

    private fun buildCommand(): List<String> {
        val socketArgs = if (cfg.socketPath.isNotEmpty()) {
            listOf("--socket-path", cfg.socketPath)
        } else {
            listOf("--port", cfg.port.toString())
        }

        val (executable, prefixArgs) = resolveExecutable()

        val serverArgs = listOf("server", "start", "--foreground", "--api-only") +
                socketArgs + cfg.args

        return listOf(executable) + prefixArgs + serverArgs
    }

    /**
     * Resolve the executable to spawn, mirroring the Go/Java fallback:
     *   1. An absolute/relative path is used directly.
     *   2. A binary name found on PATH is used as-is.
     *   3. Otherwise fall back to python/python3 (`-m orb`) if available.
     *   4. If nothing is found, fail loud with a clear error.
     *
     * @return the executable and any prefix args (e.g. `-m orb` for python).
     */
    private fun resolveExecutable(): Pair<String, List<String>> {
        val binary = cfg.binary.ifEmpty { "orb" }

        // Explicit path — trust the caller.
        if (binary.startsWith("/") || binary.startsWith(".")) {
            val prefix = if (isPythonBinary(binary)) listOf("-m", "orb") else emptyList()
            return binary to prefix
        }

        // A python interpreter passed by name runs `-m orb`.
        if (isPythonBinary(binary)) {
            return binary to listOf("-m", "orb")
        }

        // Binary name resolvable on PATH — use as-is.
        if (isOnPath(binary)) {
            return binary to emptyList()
        }

        // Fall back to python/python3 -m orb.
        for (py in listOf("python3", "python")) {
            if (isOnPath(py)) {
                return py to listOf("-m", "orb")
            }
        }

        throw IllegalStateException(
            "Cannot find ORB binary '$binary' in PATH and python/python3 not available"
        )
    }

    private fun isPythonBinary(binary: String): Boolean =
        binary.endsWith("python") || binary.endsWith("python3") || binary.endsWith("python3.exe")

    private fun isOnPath(binary: String): Boolean {
        val path = System.getenv("PATH") ?: return false
        return path.split(File.pathSeparator).any { dir ->
            File(dir, binary).canExecute()
        }
    }

    private fun buildEnvironment(): Map<String, String> {
        val env = ProcessBuilder().environment().toMutableMap()
        env.putAll(cfg.env)
        if (cfg.pythonPath.isNotEmpty()) {
            env["PYTHONPATH"] = cfg.pythonPath
        }
        return env
    }

    private fun healthUrl(): String {
        return if (cfg.socketPath.isNotEmpty()) {
            "http://localhost/health"
        } else {
            "http://localhost:${cfg.port}/health"
        }
    }

    private fun makeHealthClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .connectTimeout(Duration.ofMillis(BG_POLL_TIMEOUT_MS))
            .readTimeout(Duration.ofMillis(BG_POLL_TIMEOUT_MS))
            .callTimeout(Duration.ofMillis(BG_POLL_TIMEOUT_MS))

        if (cfg.socketPath.isNotEmpty()) {
            builder.socketFactory(UdsSocketFactory(cfg.socketPath))
        }
        return builder.build()
    }

    private fun pollHealth(): Boolean {
        return try {
            val client = makeHealthClient()
            val req = Request.Builder().url(healthUrl()).get().build()
            client.newCall(req).execute().use { resp ->
                if (resp.code == 401) return false // auth enabled but /health open
                if (resp.code != 200) return false
                val body = resp.body?.string() ?: return false
                body.contains("\"healthy\"") || body.contains("\"degraded\"")
            }
        } catch (_: Exception) {
            false
        }
    }

    private fun kill() {
        val process = lock.withLock { proc } ?: return
        val descendants = process.descendants().toList()
        process.destroyForcibly()
        descendants.forEach { it.destroyForcibly() }
    }

    private fun startBgMonitor() {
        bgJob = CoroutineScope(Dispatchers.IO).launch {
            while (isActive && !stopped.get()) {
                delay(BG_POLL_INTERVAL_MS)
                if (stopped.get()) break
                val ok = pollHealth()
                if (ok) {
                    consecutiveFail.set(0)
                    _healthy.set(true)
                } else {
                    val fails = consecutiveFail.incrementAndGet()
                    if (fails >= UNHEALTHY_THRESHOLD) {
                        _healthy.set(false)
                    }
                }
            }
        }
    }
}

/**
 * Generate a unique temp socket path for a managed ORB process.
 */
fun tempSocketPath(): String =
    "${System.getProperty("java.io.tmpdir")}/orb-${ProcessHandle.current().pid()}-${System.currentTimeMillis()}.sock"
